src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,219 @@
+  circulant (v - w) = circulant v - circulant w :=
+by ext; simp
+
+lemma circulant_mul [semiring α] [fintype I] [add_comm_group I] (v w : I → α) :
+  circulant v ⬝ circulant w = circulant (mul_vec (circulant v) w) :=
+begin
+  ext i j,
+  simp only [mul_apply, mul_vec, circulant, dot_product],
+  refine fintype.sum_equiv (equiv.sub_right j) _ _ _,
+  intro x,
+  simp only [equiv.sub_right_apply],
+  congr' 2,
+  abel
+end
+
+lemma fin.circulant_mul [semiring α] :
+  ∀ {n} (v w : fin n → α), circulant v ⬝ circulant w = circulant (mul_vec (circulant v) w)
+| 0     := dec_trivial
+| (n+1) := circulant_mul
+
+/-- multiplication of circulant matrices commutes when the elements do. -/
+lemma circulant_mul_comm
+  [comm_semigroup α] [add_comm_monoid α] [fintype I] [add_comm_group I] (v w : I → α) :
+  circulant v ⬝ circulant w = circulant w ⬝ circulant v :=
+begin
+  ext i j,
+  simp only [mul_apply, circulant, mul_comm],
+  refine fintype.sum_equiv ((equiv.sub_left i).trans (equiv.add_right j)) _ _ _,
+  intro x,
+  congr' 2,
+  { simp },
+  { simp only [equiv.coe_add_right, function.comp_app,
+               equiv.coe_trans, equiv.sub_left_apply],
+    abel }
+end
+
+lemma fin.circulant_mul_comm [comm_semigroup α] [add_comm_monoid α] :
+  ∀ {n} (v w : fin n → α), circulant v ⬝ circulant w = circulant w ⬝ circulant v
+| 0     := dec_trivial
+| (n+1) := circulant_mul_comm
+
+/-- `k • circulant v` is another circulantcluant matrix `circulant (k • v)`. -/
+lemma circulant_smul [has_sub I] [has_scalar R α] (k : R) (v : I → α) :
+  circulant (k • v) = k • circulant v :=
+by ext; simp
+
+@[simp] lemma circulant_zero (α I) [has_zero α] [has_sub I] :
+  circulant 0 = (0 : matrix I I α) :=
+by ext; simp
+
+@[simp] lemma circulant_single_one (α I) [has_zero α] [has_one α] [decidable_eq I] [add_group I] :
+  circulant (pi.single 0 1 : I → α) = (1 : matrix I I α) :=
+by { ext i j, simp only [circulant, one_apply, pi.single_apply, sub_eq_zero], congr }
+
+lemma circulant_single [semiring α] (I) [decidable_eq I] [add_group I] [fintype I] (a : α) :
+  circulant (pi.single 0 a : I → α) = scalar I a :=
+begin
+  ext i j,
+  simp [pi.single_apply, one_apply, sub_eq_zero],
+end
+
+/-- Any identity matrix is a circulant matrix. -/
+lemma one_eq_circulant (α I) [has_zero α] [has_one α] [decidable_eq I] [add_group I] :
+  (1 : matrix I I α) = circulant (λ i, ite (i = 0) 1 0) :=
+by { rw [←circulant_single_one], congr, ext i, simp [pi.single, update] }
+
+/-- Another version of `one_eq_circulant`. -/
+lemma one_eq_circulant' (α I) [has_zero α] [has_one α] [decidable_eq I] [add_group I] :
+  (1 : matrix I I α) = circulant (λ i, (1 : matrix I I α) i 0) :=
+by { ext i j, simp only [circulant, one_apply, sub_eq_zero], congr }
+
+lemma fin.one_eq_circulant (α) [has_zero α] [has_one α] :
+  ∀ n, (1 : matrix (fin n) (fin n) α) = circulant (λ i, ite (↑i = 0) 1 0)
+| 0     := dec_trivial
+| (n+1) := by simp [one_eq_circulant α (fin (n + 1)), fin.ext_iff]
```

Comment:
I think only the last of these is worth keeping - the first two are trivial consequences of how `pi.single` is defined
```suggestion
/-- Note we use `↑i = 0` instead of `i = 0` as `fin 0` has no `0`. This means that we cannot state this with
`pi.single` as we did with `matrix.circulant_single`. -/
lemma fin.circulant_ite (α) [has_zero α] [has_one α] :
  ∀ n, circulant (λ i, ite (↑i = 0) 1 0 : fin n → α) = 1
| 0     := dec_trivial
| (n+1) := begin
  simp_rw [fin.coe_eq_zero, ←circulant_single_one],
  congr' with j,
  rw pi.single_apply,
  congr,
end
```
which needs
```lean
lemma fin.coe_eq_zero {k : ℕ} (x : fin k.succ) : ↑x = 0 ↔ x = 0 :=
by rw [←fin.coe_injective.eq_iff, fin.coe_zero]
```
in `data/fin.lean`.